File: src/main/js/y2020/Day4.js
const fs = require('fs');

const input = fs.readFileSync('../../inputs/2020/4.txt', 'utf8').split('\n\n');

const parsed = input.map(row => {
  const obj = {}
  row
    .split(/[ \n]/)
    .map(cell => {
    const [a, b] = cell.split(':');
    obj[a] = b;
  });
  return obj;
});

function validateNumber(numberAsString, min, max) {
  const num = parseInt(numberAsString);
  return num.toString(10) === numberAsString && num >= min && num <= max;
}

function validateHeight(hgt) {
  if (hgt.endsWith('in')) {
    return validateNumber(hgt.substring(0, hgt.length - 2), 59, 76);
  } else if (hgt.endsWith('cm')) {
    return validateNumber(hgt.substring(0, hgt.length - 2), 150, 193);
  }
}

function validateRgb(hcl) {
  return hcl.match(/^#[0-9a-f]{6}$/g);
}

function validateIn(query, strings) {
  return strings.indexOf(query) > -1;
}

function validatePassport(pid) {
  return pid.match(/^[0-9]{9}$/g);
}

const valid = parsed.filter(item => {
  const expectedFields = ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid'];

  return expectedFields.every(field => Object.keys(item).indexOf(field) > -1) &&
    validateNumber(item.byr, 1920, 2002)
    && validateNumber(item.iyr, 2010, 2020)
    && validateNumber(item.eyr, 2020, 2030)
    && validateHeight(item.hgt)
    && validateRgb(item.hcl)
    && validateIn(item.ecl, ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'])
    && validatePassport(item.pid);
});

console.log(parsed.length); //  259
console.log(valid.length);  //  101

File: src/main/go/day4.go
package main

import (
	"fmt"
	"regexp"
	"strings"
)

func day4() {
	input := MustReadFile("../inputs/2020/4.txt")

	rows := strings.Split(input, "\n\n")

	var passports []Passport

	for _, row := range rows {
		passport, _ := parsePassport(row)
		if passport != nil {
			passports = append(passports, *passport)
		}
	}

	valid := 0

	for _, passport := range passports {
		if validatePassport(passport) {
			valid++
		}
	}

	fmt.Println(len(passports))
	fmt.Println(valid)
}

func splitByNewLineOrComma(r rune) bool {
	return r == '\n' || r == ' '
}

func validateHeight(height Height) bool {
	result := false
	if height.units == inch {
		result = numberInRange(height.count, 59, 76)
	} else if height.units == cm {
		result = numberInRange(height.count, 150, 193)
	}
	return result
}

func validatePassport(passport Passport) bool {
	return numberInRange(passport.byr, 1920, 2002) &&
		numberInRange(passport.iyr, 2010, 2020) &&
		numberInRange(passport.eyr, 2020, 2030) &&
		validateHeight(passport.hgt) &&
		validatePassportId(passport.pid)
}

func validatePassportId(pid string) bool {
	return regexp.MustCompile("^[0-9]{9}$").MatchString(pid)
}

func numberInRange(num int, min int, max int) bool {
	return num >= min && num <= max
}
